File: clients/windows/src/ClipShare.Windows.Vault/VaultMemoryGuard.cs
namespace ClipShare.Windows.Vault;

public sealed class VaultMemoryGuard
{
    private static readonly TimeSpan DefaultBackgroundLimit = TimeSpan.FromMinutes(10);
    private readonly Action clearSensitiveState;
    private readonly TimeSpan backgroundLimit;
    private DateTimeOffset? backgroundSince;

    public VaultMemoryGuard(Action clearSensitiveState, TimeSpan? backgroundLimit = null)
    {
        ArgumentNullException.ThrowIfNull(clearSensitiveState);
        this.clearSensitiveState = clearSensitiveState;
        this.backgroundLimit = backgroundLimit ?? DefaultBackgroundLimit;
        if (this.backgroundLimit < TimeSpan.Zero)
        {
            throw new ArgumentOutOfRangeException(nameof(backgroundLimit));
        }
    }

    public void OnBackground(DateTimeOffset now) => backgroundSince = now;

    public void OnForeground(DateTimeOffset now)
    {
        DateTimeOffset? started = backgroundSince;
        backgroundSince = null;
        if (started is not null && now >= started.Value + backgroundLimit)
        {
            clearSensitiveState();
        }
    }

    public void OnSystemLocked()
    {
        backgroundSince = null;
        clearSensitiveState();
    }

    public void OnProcessExit()
    {
        backgroundSince = null;
        clearSensitiveState();
    }
}

File: clients/android/core/crypto/src/main/kotlin/com/clipshare/core/crypto/VaultMemoryGuard.kt
package com.clipshare.core.crypto

import java.time.Duration
import java.time.Instant

class VaultMemoryGuard(
    private val clearSensitiveState: () -> Unit,
    private val backgroundLimit: Duration = Duration.ofMinutes(DEFAULT_BACKGROUND_MINUTES),
) {
    private var backgroundSince: Instant? = null

    init {
        require(!backgroundLimit.isNegative) { "The background limit cannot be negative." }
    }

    fun onBackground(now: Instant) {
        backgroundSince = now
    }

    fun onForeground(now: Instant) {
        val started = backgroundSince
        backgroundSince = null
        if (started != null && !now.isBefore(started.plus(backgroundLimit))) clearSensitiveState()
    }

    fun onSystemLocked() {
        backgroundSince = null
        clearSensitiveState()
    }

    fun onProcessExit() {
        backgroundSince = null
        clearSensitiveState()
    }

    private companion object {
        const val DEFAULT_BACKGROUND_MINUTES = 10L
    }
}
